bridges/snowbridge/parachain/primitives/core/src/outbound.rs

```diff
@@ -0,0 +1,408 @@
+pub trait SendMessageFeeProvider {
+	type Balance: BaseArithmetic + Unsigned + Copy;
+
+	/// The local component of the message processing fees in native currency
+	fn local_fee() -> Self::Balance;
+}
+
+/// Reasons why sending to Ethereum could not be initiated
+#[derive(Copy, Clone, Encode, Decode, PartialEq, Eq, RuntimeDebug, PalletError, TypeInfo)]
+pub enum SendError {
+	/// Message is too large to be safely executed on Ethereum
+	MessageTooLarge,
+	/// The bridge has been halted for maintenance
+	Halted,
+	/// Invalid Channel
+	InvalidChannel,
+}
+
+pub trait GasMeter {
+	/// All the gas used for submitting a message to Ethereum, minus the cost of dispatching
+	/// the command within the message
+	const MAXIMUM_BASE_GAS: u64;
+
+	fn maximum_gas_used_at_most(command: &Command) -> u64 {
+		Self::MAXIMUM_BASE_GAS + Self::maximum_dispatch_gas_used_at_most(command)
+	}
+
+	/// Measures the maximum amount of gas a command payload will require to dispatch, AFTER
+	/// validation & verification.
+	fn maximum_dispatch_gas_used_at_most(command: &Command) -> u64;
+}
+
+/// A meter that assigns a constant amount of gas for the execution of a command
+///
+/// The gas figures are extracted from this report:
+/// > forge test --match-path test/Gateway.t.sol --gas-report
+///
+/// A healthy buffer is added on top of these figures to account for:
+/// * The EIP-150 63/64 rule
+/// * Future EVM upgrades that may increase gas cost
+pub struct ConstantGasMeter;
+
+impl GasMeter for ConstantGasMeter {
+	// The base transaction cost, which includes:
+	// 21_000 transaction cost, roughly worst case 64_000 for calldata, and 100_000
+	// for message verification
+	const MAXIMUM_BASE_GAS: u64 = 185_000;
+
+	fn maximum_dispatch_gas_used_at_most(command: &Command) -> u64 {
+		match command {
+			Command::CreateAgent { .. } => 275_000,
+			Command::CreateChannel { .. } => 100_000,
+			Command::UpdateChannel { .. } => 50_000,
+			Command::TransferNativeFromAgent { .. } => 60_000,
+			Command::SetOperatingMode { .. } => 40_000,
+			Command::AgentExecute { command, .. } => match command {
+				// Execute IERC20.transferFrom
+				//
+				// Worst-case assumptions are important:
+				// * No gas refund for clearing storage slot of source account in ERC20 contract
+				// * Assume dest account in ERC20 contract does not yet have a storage slot
+				// * ERC20.transferFrom possibly does other business logic besides updating balances
+				AgentExecuteCommand::TransferToken { .. } => 100_000,
+			},
+			Command::Upgrade { initializer, .. } => {
+				let initializer_max_gas = match *initializer {
+					Some(Initializer { maximum_required_gas, .. }) => maximum_required_gas,
+					None => 0,
+				};
+				// total maximum gas must also include the gas used for updating the proxy before
+				// the the initializer is called.
+				50_000 + initializer_max_gas
+			},
+			Command::SetTokenTransferFees { .. } => 60_000,
+			Command::SetPricingParameters { .. } => 60_000,
+		}
+	}
+}
```

Comment:
nit: the numbers I'm getting locally, compared with your buffered values (see buffer calculations inline to the right of the table):
```
| test/mocks/GatewayMock.sol:GatewayMock contract |                 |        |        |        |         |
|-------------------------------------------------|-----------------|--------|--------|--------|---------|
| Deployment Cost                                 | Deployment Size |        |        |        |         |
| 4162081                                         | 21220           |        |        |        |         |
| Function Name                                   | min             | avg    | median | max    | # calls |
| agentExecute                                    | 597             | 11745  | 3471   | 39444  | 4       | -> ~60k buffer (for worst case no gas refund and no instiated storage slot?)
| agentExecutePublic                              | 4750            | 16797  | 4818   | 40823  | 3       |
| agentOf                                         | 804             | 828    | 804    | 2804   | 81      |
| channelNoncesOf                                 | 726             | 726    | 726    | 726    | 1       |
| channelOperatingModeOf                          | 568             | 2568   | 2568   | 4568   | 2       |
| createAgent                                     | 971             | 190249 | 237519 | 237519 | 10      | -> ~40k buffer
| createAgentPublic                               | 3071            | 192015 | 239252 | 239252 | 5       |
| createChannel                                   | 484             | 11727  | 2942   | 27008  | 5       | -> ~75k buffer
| createChannelPublic                             | 2835            | 16192  | 16629  | 28677  | 4       |
| implementation                                  | 664             | 664    | 664    | 664    | 1       |
| initialize                                      | 2749            | 625106 | 640665 | 640665 | 41      |
| operatingMode                                   | 608             | 1608   | 1608   | 2608   | 4       |
| pricingParameters                               | 4430            | 4430   | 4430   | 4430   | 1       |
| quoteRegisterTokenFee                           | 11872           | 20622  | 22372  | 22372  | 12      |
| quoteSendTokenFee                               | 12551           | 12588  | 12601  | 12601  | 4       |
| registerToken                                   | 6506            | 75532  | 79867  | 99873  | 10      |
| sendToken                                       | 34880           | 44787  | 48408  | 52300  | 7       |
| setCommitmentsAreVerified                       | 4272            | 21995  | 22439  | 22439  | 41      |
| setOperatingMode                                | 836             | 3402   | 2444   | 24344  | 44      | -> ~15k buffer
| setOperatingModePublic                          | 4199            | 5217   | 4199   | 26099  | 43      |
| setPricingParameters                            | 4532            | 4532   | 4532   | 4532   | 1       | -> ~55k buffer
| setPricingParametersPublic                      | 6387            | 6387   | 6387   | 6387   | 1       |
| setTokenTransferFees                            | 4779            | 4779   | 4779   | 4779   | 1       |
| setTokenTransferFeesPublic                      | 6316            | 6316   | 6316   | 6316   | 1       |
| submitV1                                        | 1756            | 152025 | 162409 | 286134 | 8       |
| transferNativeFromAgent                         | 924             | 22795  | 22795  | 44667  | 2       | -> ~15k buffer
| transferNativeFromAgentPublic                   | 46116           | 46116  | 46116  | 46116  | 1       |
| updateChannel                                   | 946             | 7510   | 3383   | 26783  | 5       | -> ~25k buffer
| updateChannelPublic                             | 4782            | 11035  | 5339   | 28682  | 4       |
| upgrade                                         | 509             | 9195   | 6135   | 29122  | 5       | -> ~25k+ buffer
| upgradePublic                                   | 3251            | 12954  | 8924   | 30717  | 4       |
```

Some of these gas figures have changed quite a bit since my prior snowfork checkout
(a22232958570cdd1983b042ea37f78fbdd3eb455), so checking in whether the numbers here are still up to date, or whether there's another reason these buffers differ wildly. EIP-150 63/64 should affect the buffering equally.
In any case, it would be nice to document how these buffers were determined, to make it more
efficient to maintain healthy buffers in future.

Reply:
Yes, agree, I haven't really being following a deterministic process for calculating those buffers.

I also increased the buffer for more expensive commands like `CreateAgent`, relative to less expensive commands, just as an extra measure of relief. But I guess there isn't really a good reason to do that, so will revert to just making the buffer 20% for all commands.